File: apps/api/app/billing/razorpay.py
import hashlib
import hmac
from typing import cast

import httpx


class RazorpayError(Exception):
	pass


class RazorpayUnavailableError(RazorpayError):
	pass


def verify_checkout_signature(
	order_id: str, payment_id: str, signature: str, key_secret: str
) -> bool:
	expected = hmac.new(
		key_secret.encode(), f"{order_id}|{payment_id}".encode(), hashlib.sha256
	).hexdigest()
	return hmac.compare_digest(expected, signature)


def verify_webhook_signature(body: bytes, signature: str, webhook_secret: str) -> bool:
	expected = hmac.new(webhook_secret.encode(), body, hashlib.sha256).hexdigest()
	return hmac.compare_digest(expected, signature)


class RazorpayClient:
	def __init__(
		self,
		key_id: str,
		key_secret: str,
		base_url: str = "https://api.razorpay.com/v1",
		timeout_seconds: float = 15.0,
		transport: httpx.AsyncBaseTransport | None = None,
	) -> None:
		if not key_id or not key_secret:
			raise RazorpayUnavailableError("Razorpay credentials are not configured")
		self._auth = (key_id, key_secret)
		self._base_url = base_url.rstrip("/")
		self._timeout_seconds = timeout_seconds
		self._transport = transport

	async def create_order(
		self,
		amount_paise: int,
		currency: str,
		receipt: str,
		notes: dict[str, str] | None = None,
	) -> dict[str, object]:
		body = await self._request(
			"POST",
			"/orders",
			json={
				"amount": amount_paise,
				"currency": currency,
				"receipt": receipt,
				"notes": notes or {},
			},
		)
		if not isinstance(body.get("id"), str):
			raise RazorpayError("Razorpay order response is missing an identifier")
		return body

	async def order_payments(self, razorpay_order_id: str) -> list[dict[str, object]]:
		body = await self._request("GET", f"/orders/{razorpay_order_id}/payments")
		items = body.get("items")
		if not isinstance(items, list):
			raise RazorpayError("Razorpay payments response is malformed")
		return [item for item in items if isinstance(item, dict)]

	async def _request(self, method: str, path: str, **kwargs: object) -> dict[str, object]:
		try:
			async with httpx.AsyncClient(
				timeout=self._timeout_seconds,
				transport=self._transport,
				auth=self._auth,
			) as client:
				response = await client.request(method, self._base_url + path, **kwargs)  # type: ignore[arg-type]
		except httpx.HTTPError as error:
			raise RazorpayUnavailableError("Razorpay request failed") from error
		try:
			body = response.json()
		except ValueError as error:
			raise RazorpayError(
				f"Razorpay returned invalid JSON ({response.status_code})"
			) from error
		if not isinstance(body, dict):
			raise RazorpayError("Razorpay returned an unexpected response shape")
		if response.status_code >= 400:
			description = ""
			error = body.get("error")
			if isinstance(error, dict):
				description = str(cast(dict[str, object], error).get("description", ""))
			raise RazorpayError(description or f"Razorpay returned HTTP {response.status_code}")
		return body


def payment_entity(payload: dict[str, object]) -> dict[str, object]:
	entity = payload.get("payload")
	if isinstance(entity, dict):
		payment = entity.get("payment")
		if isinstance(payment, dict):
			inner = cast(dict[str, object], payment).get("entity")
			if isinstance(inner, dict):
				return inner
	return {}


def refund_entities(payment: dict[str, object]) -> list[dict[str, object]]:
	refunds = payment.get("refunds")
	if not isinstance(refunds, list):
		return []
	return [refund for refund in refunds if isinstance(refund, dict)]

File: apps/api-new/src/billing/razorpay.ts
import { createHmac, timingSafeEqual } from "node:crypto";

export class RazorpayError extends Error {}

export class RazorpayUnavailableError extends RazorpayError {}

const signaturesEqual = (expected: string, actual: string): boolean => {
	const expectedBytes = Buffer.from(expected, "utf8");
	const actualBytes = Buffer.from(actual, "utf8");
	return expectedBytes.length === actualBytes.length && timingSafeEqual(expectedBytes, actualBytes);
};

export const verifyCheckoutSignature = (orderId: string, paymentId: string, signature: string, keySecret: string): boolean => {
	const expected = createHmac("sha256", keySecret).update(`${orderId}|${paymentId}`).digest("hex");
	return signaturesEqual(expected, signature);
};

export const verifyWebhookSignature = (body: Uint8Array, signature: string, webhookSecret: string): boolean => {
	const expected = createHmac("sha256", webhookSecret).update(body).digest("hex");
	return signaturesEqual(expected, signature);
};

type RazorpayBody = Record<string, unknown>;

const isObject = (value: unknown): value is RazorpayBody => typeof value === "object" && value !== null && !Array.isArray(value);

export class RazorpayClient {
	private readonly authorization: string;
	private readonly baseUrl: string;
	private readonly timeoutMilliseconds: number;

	public constructor(keyId: string, keySecret: string, baseUrl = "https://api.razorpay.com/v1", timeoutSeconds = 15) {
		if (!keyId || !keySecret) throw new RazorpayUnavailableError("Razorpay credentials are not configured");
		this.authorization = `Basic ${Buffer.from(`${keyId}:${keySecret}`).toString("base64")}`;
		this.baseUrl = baseUrl.replace(/\/+$/, "");
		this.timeoutMilliseconds = timeoutSeconds * 1000;
	}

	public async createOrder(amountPaise: number, currency: string, receipt: string, notes?: Record<string, string>): Promise<RazorpayBody> {
		const body = await this.request("POST", "/orders", {
			amount: amountPaise,
			currency,
			receipt,
			notes: notes ?? {},
		});
		if (typeof body["id"] !== "string") throw new RazorpayError("Razorpay order response is missing an identifier");
		return body;
	}

	public async orderPayments(razorpayOrderId: string): Promise<RazorpayBody[]> {
		const body = await this.request("GET", `/orders/${razorpayOrderId}/payments`);
		if (!Array.isArray(body["items"])) throw new RazorpayError("Razorpay payments response is malformed");
		return body["items"].filter(isObject);
	}

	private async request(method: string, path: string, requestBody?: RazorpayBody): Promise<RazorpayBody> {
		const controller = new AbortController();
		const timeout = setTimeout(() => controller.abort(), this.timeoutMilliseconds);
		let response: Response;
		try {
			response = await fetch(`${this.baseUrl}${path}`, {
				method,
				headers: { Authorization: this.authorization, ...(requestBody ? { "Content-Type": "application/json" } : {}) },
				...(requestBody ? { body: JSON.stringify(requestBody) } : {}),
				signal: controller.signal,
			});
		} catch (cause) {
			throw new RazorpayUnavailableError("Razorpay request failed", { cause });
		} finally {
			clearTimeout(timeout);
		}

		let body: unknown;
		try {
			body = await response.json();
		} catch (cause) {
			throw new RazorpayError(`Razorpay returned invalid JSON (${response.status})`, { cause });
		}
		if (!isObject(body)) throw new RazorpayError("Razorpay returned an unexpected response shape");
		if (!response.ok) {
			const error = body["error"];
			const description = isObject(error) && typeof error["description"] === "string" ? error["description"] : "";
			throw new RazorpayError(description || `Razorpay returned HTTP ${response.status}`);
		}
		return body;
	}
}

export const paymentEntity = (payload: RazorpayBody): RazorpayBody => {
	const payloadValue = payload["payload"];
	const paymentValue = isObject(payloadValue) ? payloadValue["payment"] : undefined;
	return isObject(paymentValue) && isObject(paymentValue["entity"]) ? paymentValue["entity"] : {};
};

export const refundEntities = (payment: RazorpayBody): RazorpayBody[] => Array.isArray(payment["refunds"]) ? payment["refunds"].filter(isObject) : [];
